src/pages/signin/SignInPage.js

```diff
@@ -33,149 +30,157 @@ const propTypes = {
 
+    useEffect(() => Performance.measureTTI(), []);
+    useEffect(() => {
         App.setLocale(Localize.getDevicePreferredLocale());
-    }
-
-    render() {
-        // Show the login form if
-        // - A login has not been entered yet
-        // - AND a validateCode has not been cached with sign in link
-        const showLoginForm = !this.props.credentials.login && !this.props.credentials.validateCode;
-
-        // Show the unlink form if
-        // - A login has been entered
-        // - AND the login is not the primary login
-        // - AND the login is not validated
-        const showUnlinkLoginForm =
-            this.props.credentials.login && this.props.account.primaryLogin && this.props.account.primaryLogin !== this.props.credentials.login && !this.props.account.validated;
-
-        // Show the old password form if
-        // - A login has been entered
-        // - AND an account exists and is validated for this login
-        // - AND a password hasn't been entered yet
-        // - AND haven't forgotten password
-        // - AND the login isn't an unvalidated secondary login
-        // - AND the user is NOT on the passwordless beta
-        const showPasswordForm =
-            Boolean(this.props.credentials.login) &&
-            this.props.account.validated &&
-            !this.props.credentials.password &&
-            !this.props.account.forgotPassword &&
-            !showUnlinkLoginForm &&
-            !Permissions.canUsePasswordlessLogins(this.props.betas);
-
-        // Show the new magic code / validate code form if
-        // - A login has been entered or a validateCode has been cached from sign in link
-        // - AND the login isn't an unvalidated secondary login
-        // - AND the user is on the 'passwordless' beta
-        const showValidateCodeForm = (this.props.credentials.login || this.props.credentials.validateCode) && !showUnlinkLoginForm && Permissions.canUsePasswordlessLogins(this.props.betas);
-
-        // Show the resend validation link form if
-        // - A login has been entered
-        // - AND is not validated or password is forgotten
-        // - AND the login isn't an unvalidated secondary login
-        // - AND user is not on 'passwordless' beta
-        const showResendValidationForm =
-            Boolean(this.props.credentials.login) &&
-            (!this.props.account.validated || this.props.account.forgotPassword) &&
-            !showUnlinkLoginForm &&
-            !Permissions.canUsePasswordlessLogins(this.props.betas);
-
-        let welcomeHeader = '';
-        let welcomeText = '';
-        if (showValidateCodeForm) {
-            if (this.props.account.requiresTwoFactorAuth) {
+    }, []);
+
+    /*
+     * Show the login form if:
+     *   - A login has not been entered yet
+     *   - AND a validateCode has not been cached with the sign in link
+     */
+    const shouldShowLoginForm = !credentials.login && !credentials.validateCode;
```

Comment:
Wouldn't be better to move the code from line 64-106, (all the flags creation) into a function outside of the component? so at the en will looks like something like:

```
const { 
      shouldShowLoginForm,
       shouldShowValidateCodeForm,
        shouldShowPasswordForm,
        shouldShowUnlinkLoginForm,
        shouldShowResendValidationForm} = getLoginFormStatus(credentials, account);   
```
This will encapsulate the logic, make the functional component more readable, will hide middle flags that aren't needed in the component like `isUnvalidatedSecondaryLogin`

Reply:
I was unsure about whether I agreed with this change, but I think I implemented it in such a way that the code became more self-documenting and I could remove the comments because they weren't really adding any additional clarity, so that was a win 👍🏼